File: core/FindAndReplacePattern/FindAndReplacePattern.java
package com.dbc;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class FindAndReplacePattern {
    public List<String> findAndReplacePattern(String[] words, String pattern) {
        List<String> res = new ArrayList<>();
        for (String word : words) {
            if (word.length() != pattern.length()) continue;
            Map<Character, Character> map1 = new HashMap<>();
            Map<Character, Character> map2 = new HashMap<>();
            for (int i = 0; i < word.length(); i++) {
                if (!map1.containsKey(word.charAt(i)) && !map2.containsKey(pattern.charAt(i))) {
                    map1.put(word.charAt(i), pattern.charAt(i));
                    map2.put(pattern.charAt(i), word.charAt(i));
                }
                if (map1.getOrDefault(word.charAt(i), '_') != pattern.charAt(i) || map2.getOrDefault(pattern.charAt(i), '_') != word.charAt(i)) break;
                if (i == word.length() - 1) res.add(word);
            }
        }
        return res;
    }
}

File: core/FindAndReplacePattern/FindAndReplacePattern.py
from typing import List


def findAndReplacePattern(words: List[str], pattern: str) -> List[str]:
    res = []

    for word in words:
        if len(word) != len(pattern):
            continue
        mapping1, mapping2, flag = {}, {}, True
        for ch1, ch2 in zip(word, pattern):
            if ch1 not in mapping1 and ch2 not in mapping2:
                mapping1[ch1] = ch2
                mapping2[ch2] = ch1
            if mapping1.get(ch1, "") != ch2 or mapping2.get(ch2, "") != ch1:
                flag = False
                break

        if flag:
            res.append(word)

    return res





